Load all necessary Libraries

---



In [ ]:
import numpy as np
import pandas as pd
import math
import re
from nltk import sent_tokenize, word_tokenize, PorterStemmer
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from keras.preprocessing.text import Tokenizer
from keras.preprocessing.sequence import pad_sequences
import keras
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score

In [ ]:
df = pd.read_csv("/content/Sentiment.csv")
df.head()

id               candidate  candidate_confidence relevant_yn  \
0   1  No candidate mentioned                   1.0         yes   
1   2            Scott Walker                   1.0         yes   
2   3  No candidate mentioned                   1.0         yes   
3   4  No candidate mentioned                   1.0         yes   
4   5            Donald Trump                   1.0         yes   

   relevant_yn_confidence sentiment  sentiment_confidence     subject_matter  \
0                     1.0   Neutral                0.6578  None of the above   
1                     1.0  Positive                0.6333  None of the above   
2                     1.0   Neutral                0.6629  None of the above   
3                     1.0  Positive                1.0000  None of the above   
4                     1.0  Positive                0.7045  None of the above   

   subject_matter_confidence candidate_gold  ... relevant_yn_gold  \
0                     1.0000            NaN  ...              NaN   
1                     1.0000            NaN  ...              NaN   
2                     0.6629            NaN  ...              NaN   
3                     0.7039            NaN  ...              NaN   
4                     1.0000            NaN  ...              NaN   

  retweet_count  sentiment_gold subject_matter_gold  \
0             5             NaN                 NaN   
1            26             NaN                 NaN   
2            27             NaN                 NaN   
3           138             NaN                 NaN   
4           156             NaN                 NaN   

                                                text tweet_coord  \
0  RT @NancyLeeGrahn: How did everyone feel about...         NaN   
1  RT @ScottWalker: Didn't catch the full #GOPdeb...         NaN   
2  RT @TJMShow: No mention of Tamir Rice and the ...         NaN   
3  RT @RobGeorge: That Carly Fiorina is trending ...         NaN   
4  RT @DanScavino: #GOPDebate w/ @realDonaldTrump...         NaN   

               tweet_created            tweet_id  tweet_location  \
0  2015-08-07 09:54:46 -0700  629697200650592256             NaN   
1  2015-08-07 09:54:46 -0700  629697199560069120             NaN   
2  2015-08-07 09:54:46 -0700  629697199312482304             NaN   
3  2015-08-07 09:54:45 -0700  629697197118861312           Texas   
4  2015-08-07 09:54:45 -0700  629697196967903232             NaN   

                user_timezone  
0                       Quito  
1                         NaN  
2                         NaN  
3  Central Time (US & Canada)  
4                     Arizona  

[5 rows x 21 columns]

In [ ]:
df.shape

(13871, 21)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13871 entries, 0 to 13870
Data columns (total 21 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   id                         13871 non-null  int64  
 1   candidate                  13775 non-null  object 
 2   candidate_confidence       13871 non-null  float64
 3   relevant_yn                13871 non-null  object 
 4   relevant_yn_confidence     13871 non-null  float64
 5   sentiment                  13871 non-null  object 
 6   sentiment_confidence       13871 non-null  float64
 7   subject_matter             13545 non-null  object 
 8   subject_matter_confidence  13871 non-null  float64
 9   candidate_gold             28 non-null     object 
 10  name                       13871 non-null  object 
 11  relevant_yn_gold           32 non-null     object 
 12  retweet_count              13871 non-null  int64  
 13  sentiment_gold             15 non-null     obj

In [ ]:
df["sentiment"].unique()

array(['Neutral', 'Positive', 'Negative'], dtype=object)

#Q1. Total Number of Positive and Negative Sentiment

In [ ]:
p_count = 0
n_count = 0
for i in df["sentiment"]:
    if i == "Positive":
        p_count += 1
    elif i == "Negative":
        n_count += 1
print("Positive count : ", p_count)
print("Negative count : ", n_count)
print("Total positive and negative count : ", p_count + n_count)

Positive count :  2236
Negative count :  8493
Total positive and negative count :  10729


The total number of poitive and negative count is 10729



In [ ]:
df = df[df["sentiment"] != "Neutral"]

In [ ]:
df.shape

(10729, 21)

#Q2. Sequential LSTM Model

In [ ]:
df_new = df[["sentiment", "text"]]
df_new.head()


sentiment                                               text
1  Positive  RT @ScottWalker: Didn't catch the full #GOPdeb...
3  Positive  RT @RobGeorge: That Carly Fiorina is trending ...
4  Positive  RT @DanScavino: #GOPDebate w/ @realDonaldTrump...
5  Positive  RT @GregAbbott_TX: @TedCruz: "On my first day ...
6  Negative  RT @warriorwoman91: I liked her and was happy ...

Data Preprocessing

In [ ]:
def remove_tags(string):
    removelist = ""
    result = re.sub('RT','',string) # Remove RT from text
    result = result.lower()
    return result
df_new['text'] = df_new['text'].apply(lambda cw : remove_tags(cw))

<ipython-input-39-6063192e89e0>:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_new['text'] = df_new['text'].apply(lambda cw : remove_tags(cw))


In [ ]:
import nltk
nltk.download('stopwords')
stop_words = set(stopwords.words('english'))
df_new['text'] = df_new['text'].apply(lambda x: ' '.join([word for word in x.split() if word not in (stop_words)]))

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
<ipython-input-34-45a9a7a9f384>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_new['text'] = df_new['text'].apply(lambda x: ' '.join([word for word in x.split() if word not in (stop_words)]))


In [ ]:
nltk.download('wordnet')
w_tokenizer = nltk.tokenize.WhitespaceTokenizer()
lemmatizer = nltk.stem.WordNetLemmatizer()
def lemmatize_text(text):
    st = ""
    for w in w_tokenizer.tokenize(text):
        st = st + lemmatizer.lemmatize(w) + " "
    return st
df_new['text'] = df_new.text.apply(lemmatize_text)
df_new.sample(15)

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
<ipython-input-35-e5f455d45ff3>:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_new['text'] = df_new.text.apply(lemmatize_text)


sentiment                                               text
13561  Negative  @teapainusa: ted cruz speaking, america parent...
4736   Negative  glad covered #blacklivesmatter &amp; #climatec...
7961   Negative  chris christie: lip moving, he's still lying h...
2357   Negative  #gopdebate very, put politely, macho? people a...
8480   Negative  tonight learned @govchristie think remove civi...
8955   Negative  @rwsurfergirl: american people pick next presi...
2136   Negative  following @pattonoswalt last night, #gopdebate...
11158  Negative  @rwsurfergirl: anyone else want punch chris wa...
6803   Negative  @alexisinnh fox clearly forgot entire issue so...
4933   Negative  please tell @realdonaldtrump difference "hones...
279    Negative  @pattonoswalt: despite huckabee says, always m...
11571  Negative  @rwsurfergirl: @govchristie hugged obama deser...
2064   Negative  @thepatriot143: sure gop stage looked nice &am...
12551  Negative  @supermanhotmale: scott walker: business iran,...
9483   Negative  @lindasuhler: jeb stand "act love"; comment ea...

**Encoding  Labels**

In [ ]:
reviews = df_new["text"].values
labels = df_new["sentiment"].values
encoder = LabelEncoder()
encoded_labels = encoder.fit_transform(labels)

**Train Test Split**

In [ ]:
train_sentences, test_sentences, train_labels, test_labels = train_test_split(reviews, encoded_labels, stratify = encoded_labels)

 **Tokenizing Sentences**

In [ ]:
vocab_size = 3000
oov_tok = ''
embedding_dim = 100
max_length = 200
padding_type='post'
trunc_type='post'

tokenizer = Tokenizer(num_words = vocab_size, oov_token=oov_tok)
tokenizer.fit_on_texts(train_sentences)
word_index = tokenizer.word_index

train_sequences = tokenizer.texts_to_sequences(train_sentences)
train_padded = pad_sequences(train_sequences, padding='post', maxlen=max_length)

test_sequences = tokenizer.texts_to_sequences(test_sentences)
test_padded = pad_sequences(test_sequences, padding='post', maxlen=max_length)


**Building model**

In [ ]:
model = keras.Sequential([
    keras.layers.Embedding(vocab_size, embedding_dim, input_length=max_length),
    keras.layers.Bidirectional(keras.layers.LSTM(64)),
    keras.layers.Dense(24, activation='relu'),
    keras.layers.Dense(1, activation='sigmoid')
])
model.compile(loss='binary_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding (Embedding)       (None, 200, 100)          300000    
                                                                 
 bidirectional (Bidirection  (None, 128)               84480     
 al)                                                             
                                                                 
 dense (Dense)               (None, 24)                3096      
                                                                 
 dense_1 (Dense)             (None, 1)                 25        
                                                                 
Total params: 387601 (1.48 MB)
Trainable params: 387601 (1.48 MB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [ ]:
num_epochs = 5
history = model.fit(train_padded, train_labels,
                    epochs = num_epochs, verbose = 1,
                    validation_split = 0.1)

Epoch 1/5
227/227 [==============================] - 92s 360ms/step - loss: 0.4126 - accuracy: 0.8315 - val_loss: 0.3355 - val_accuracy: 0.8559
Epoch 2/5
227/227 [==============================] - 61s 271ms/step - loss: 0.2890 - accuracy: 0.8768 - val_loss: 0.3430 - val_accuracy: 0.8571
Epoch 3/5
227/227 [==============================] - 61s 267ms/step - loss: 0.2339 - accuracy: 0.9033 - val_loss: 0.3518 - val_accuracy: 0.8571
Epoch 4/5
227/227 [==============================] - 61s 270ms/step - loss: 0.1931 - accuracy: 0.9205 - val_loss: 0.4365 - val_accuracy: 0.8547
Epoch 5/5
227/227 [==============================] - 60s 264ms/step - loss: 0.1550 - accuracy: 0.9377 - val_loss: 0.4474 - val_accuracy: 0.8472


In [ ]:
prediction = model.predict(test_padded)
pred_labels = []

for i in prediction:
    if i >= 0.5:
        pred_labels.append(1)
    else:
        pred_labels.append(0)
print("Accuracy of prediction on test set : ", accuracy_score(test_labels,pred_labels))

84/84 [==============================] - 5s 54ms/step
Accuracy of prediction on test set :  0.839731643682445


#Q3. Checking the sentiment for the following two sentences
a. 'He is a great leader.'
b. 'He is a terrible leader.

In [ ]:
sentence = ["He is a great leader.",
            "He is a terrible leader."]

sequences = tokenizer.texts_to_sequences(sentence)

padded = pad_sequences(sequences, padding='post', maxlen=max_length)
prediction = model.predict(padded)
pred_labels = []

for i in prediction:
    if i >= 0.5:
        pred_labels.append(1)
    else:
        pred_labels.append(0)

for i in range(len(sentence)):
    print(sentence[i])
    if pred_labels[i] == 1:
        s = 'Positive'
    else:
        s = 'Negative'
    print("Predicted sentiment : ",s)

1/1 [==============================] - 0s 42ms/step
He is a great leader.
Predicted sentiment :  Positive
He is a terrible leader.
Predicted sentiment :  Negative
